### ISSUE TITLE
useless_vec reports non-constant repeat counts (sometimes)

### ISSUE THREAD

**Issue by AtheMathmo** [Manishearth/rust-clippy#1063]
This is happening with clippy v0.0.77 and the latest nightly build (at the time of posting).

I believe this is a bug - if not perhaps the suggestion from clippy should be changed.

The `useless_vec` is triggering on situations where I am setting the length with an expression rather than a compile time constant. It suggests I remove the `vec!` part which leads to error [E0307](https://doc.rust-lang.org/error-index.html#E0307) from the rust compiler.

The error only seems to arise if the `&vec![]` is being passed as an argument to another function.

To reproduce:

```rust
struct Line {
    length: usize
}

impl Line {
    fn length(&self) -> usize {
        self.length
    }
}

fn do_something(points: &[usize]) { }

fn main() {
    let line = Line { length: 3 };

    // This won't trigger clippy warning
    let points = &vec![1; line.length];

    // This won't trigger clippy warning
    let points = &vec![1; line.length()];

    // This will
    do_something(&vec![1, line.length]);

    // As will this
    do_something(&vec![1, line.length()]);
}
```

Clippy will complain about the last two calls - saying that `vec!` is unnecessary and should be removed. However, removing this leads to the [E0307](https://doc.rust-lang.org/error-index.html#E0307) error.

**Comment by AtheMathmo** [comment]
I saw this has been tagged easy. I'd be happy to make the change myself if you're open to providing some pointers to the right places to look.

**Comment by Manishearth** [comment]
The relevant code is in https://github.com/Manishearth/rust-clippy/blob/f3397afacbedaf8c8c8d8928b7c146ed000d3a2c/clippy_lints/src/vec.rs#L53

The `Args::Repeat` case is the one we care about; `Args::Repeat(value, times)` is clippy's internal representation of `vec![value, times]`.

`times` is an [`Expr`](http://manishearth.github.io/rust-internals-docs/rustc/hir/struct.Expr.html), to fix this bug we need to ensure that its node is an [`ExprLit`](http://manishearth.github.io/rust-internals-docs/rustc/hir/enum.Expr_.html), i.e., a literal.

Tests would need to be added to https://github.com/Manishearth/rust-clippy/blob/master/tests/compile-fail/vec.rs

**Comment by mcarton** [comment]
> times is an Expr, to fix this bug we need to ensure that its node is an ExprLit, i.e., a literal.

Any constant expression would do actually. One could check if [`eval_const_expr_partial`](https://manishearth.github.io/rust-internals-docs/rustc_const_eval/fn.eval_const_expr_partial.html) is ok.

**Comment by AtheMathmo** [comment]
I tried implementing the `eval_const_expr_partial` as @mcarton suggested. I added this line:

`let Ok(_) = eval_const_expr_partial(cx.tcx, expr, ExprTypeChecked, None),` [here](https://github.com/Manishearth/rust-clippy/blob/f3397afacbedaf8c8c8d8928b7c146ed000d3a2c/clippy_lints/src/vec.rs#L39).

It compiles but causes the existing compile-fail tests to fail. Do you mind giving another pointer? :)

**Comment by mcarton** [comment]
That `expr` is the whole `vec![…]` expression. You need to do it [here](https://github.com/Manishearth/rust-clippy/blob/f3397afacbedaf8c8c8d8928b7c146ed000d3a2c/clippy_lints/src/vec.rs#L56) on `len`.
By the way, there is a [`Result::is_ok`](https://doc.rust-lang.org/core/result/enum.Result.html#method.is_ok) method :smile:

**Comment by AtheMathmo** [comment]
Ah, I thought the place you linked was for once an offending macro had been found. So if I find here that the value is non-constant I should return what for `snippet`?

As for `is_ok` I thought I needed to use `let` within that macro - I'll use `is_ok` within the match block :)
